web/web.go

```diff
@@ -0,0 +1,1022 @@
+		return
+	}
+	if table == "" || pos == "" {
+		http.Error(w, fmt.Sprintf("Table name or position is empty"), http.StatusBadRequest)
+	}
+	sp := app.conv.SpSchema[table]
+	position, err := strconv.Atoi(pos)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Error converting position to integer"), http.StatusBadRequest)
+		return
+	}
+	if position < 0 || position >= len(sp.Indexes) {
+		http.Error(w, fmt.Sprintf("No secondary index found at position %d", position), http.StatusBadRequest)
+		return
+	}
+	sp.Indexes = removeSecondaryIndex(sp.Indexes, position)
+	app.conv.SpSchema[table] = sp
+	updateSessionFile()
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(app.conv)
+}
+
+// updateSessionFile updates the content of session file with
+// latest app.conv.
+func updateSessionFile() error {
+	filePath := app.sessionFile
+	if filePath == "" {
+		return fmt.Errorf("Session file path is empty")
+	}
+	f, err := os.OpenFile(filePath, os.O_RDWR|os.O_CREATE, 0644)
+	if err != nil {
+		return err
+	}
+	convJSON, err := json.MarshalIndent(app.conv, "", " ")
+	if err != nil {
+		return err
+	}
+	err = f.Truncate(0)
+	if err != nil {
+		return err
+	}
+	if _, err := f.Write(convJSON); err != nil {
+		return err
+	}
+	return nil
+}
+
+// rollback is used to get previous state of conversion in case
+// some unexpected error occurs during update operations.
+func rollback() error {
+	if app.sessionFile != "" {
+		app.conv = internal.MakeConv()
+		err := conversion.ReadSessionFile(app.conv, app.sessionFile)
+		if err != nil {
+			return fmt.Errorf("Failed to rollback action: %v", err)
+		}
+	}
+	return nil
+}
+
+func isPartOfPK(col, table string) bool {
+	for _, pk := range app.conv.SpSchema[table].Pks {
+		if pk.Col == col {
+			return true
+		}
+	}
+	return false
+}
+
+func isParent(table string) bool {
+	for _, spSchema := range app.conv.SpSchema {
+		if spSchema.Parent == table {
+			return true
+		}
+	}
+	return false
+}
+
+func isPartOfSecondaryIndex(col, table string) bool {
+	for _, index := range app.conv.SpSchema[table].Indexes {
+		for _, key := range index.Keys {
+			if key.Col == col {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func isPartOfFK(col, table string) bool {
+	for _, fk := range app.conv.SpSchema[table].Fks {
+		for _, column := range fk.Columns {
+			if column == col {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// TODO: create a map to store referenced column to get
+// this information in O(1).
+func isReferencedByFK(col, table string) bool {
+	for _, spSchema := range app.conv.SpSchema {
+		if table != spSchema.Name {
+			for _, fk := range spSchema.Fks {
+				if fk.ReferTable == table {
+					for _, column := range fk.ReferColumns {
+						if column == col {
+							return true
+						}
+					}
+				}
+			}
+		}
+	}
+	return false
+}
+
+func canRemoveColumn(colName, table string) (error, int) {
```

Comment:
I'm a little concerned about all of the calls to rollback in this and the following functions.

It is adding lots of code, and it is error prone: we need to make sure we add a rollback to each of the error paths we encounter on each of these functions, and we need to remember to add them to any future code.

I suggest just returning errors from these functions, and handling rollback uniformly in updateTableSchema. I'll add a comment there with a suggestion of how to structure this.

Reply:
The way to handle this in Go is to capture error that is returned in a named return variable e.g., 'func canRemoveColumn(colName, table string) (err error, code int)'. Then you can have a defer() statement that looks at the value of 'err' and if 'err' is not nil, then do the rollback.

That said, I agree with Nevin's assessment that the rollback should be done at the outermost function that populates http response which is updateTableSchema in this case.

Reply:
added TODO, will change in the following PR